File: java/leetcode/src/Task406.java
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class Task406 {
    public int[][] reconstructQueue(int[][] people) {
        if (people.length < 2){
            return people;
        }
        int[][] res = new int[people.length][2];
        //init
        for (int[] node : res) {
            node[0]=-1;
        }

        Arrays.sort(people, Comparator.comparingInt(a -> a[0]));

        for (int i = 0; i < people.length; i++) {
            int[] node = people[i];
            int j = 0,empty=0;
            while(empty != node[1]+1){
                if(res[j][0] == -1 || res[j][0] == node[0]){
                    ++empty;
                }
                ++j;
            }
            res[j-1] = node;

        }

        return res;

    }

    public int[][] reconstructQueue2(int[][] people) {
        if (people.length < 2){
            return people;
        }


        List<int[]> res = new ArrayList<>();

        Arrays.sort(people, (a,b)-> a[0] == b[0]? a[1]-b[1] : b[0]-a[0]);

        for (int[] person : people) {
            res.add(person[1],person);
        }

        return res.toArray(new int[people.length][2]);

    }
    @Test
    void test(){
        int[][] people = new int[][]{{7,0}, {4,4}, {7,1}, {5,0}, {6,1}, {5,2}};
        int[][] res = reconstructQueue2(people);
    }

}

File: py/Task406.py
from typing import List
from operator import itemgetter


class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        people.sort(key=lambda x: (x[0], -x[1]))
        n = len(people)
        res = [None] * n
        for person in people:
            i, k = -1, person[1]
            while k >= 0:
                i += 1
                if not res[i]:
                    k -= 1
            res[i] = person
        return res

    # 从大到小排序后直接插入数组
    def reconstructQueue2(self, people: List[List[int]]) -> List[List[int]]:
        people.sort(key=lambda x: (-x[0], x[1]))
        res = []
        for person in people:
            res.insert(person[1], person)
        return res


people = [[7, 0], [4, 4], [7, 1], [5, 0], [6, 1], [5, 2]]
obj = Solution()
print(obj.reconstructQueue2(people))
